test/Microsoft.Azure.ServiceBus.UnitTests/RetryTests.cs

```diff
@@ -0,0 +1,192 @@
+            ServerBusyException exception = new ServerBusyException(string.Empty);
+            TimeSpan retryInterval;
+
+            // First ServerBusy exception
+            Assert.False(policy1.IsServerBusy, "policy1.IsServerBusy should start with false");
+            Assert.True(policy1.ShouldRetry(duration, retryCount, exception, out retryInterval), "We should retry, but it returned false");
+            Assert.True(policy1.IsServerBusy, "policy1.IsServerBusy should be true");
+
+            System.Threading.Thread.Sleep(3000);
+
+            // Setting it a second time should not prolong the call.
+            Assert.True(policy1.IsServerBusy, "policy1.IsServerBusy should be true");
+            Assert.True(policy1.ShouldRetry(duration, retryCount, exception, out retryInterval), "We should retry, but it return false");
+            Assert.True(policy1.IsServerBusy, "policy1.IsServerBusy should be true");
+
+            System.Threading.Thread.Sleep(8000); // 3 + 8 = 11s
+            Assert.False(policy1.IsServerBusy, "policy1.IsServerBusy should stay false after 11s");
+
+            // Setting ServerBusy for second time.
+            Assert.True(policy1.ShouldRetry(duration, retryCount, exception, out retryInterval), "We should retry, but it return false");
+            Assert.True(policy1.IsServerBusy, "policy1.IsServerBusy is not true");
+        }
+
+        [Fact]
+        async void RunOperationShouldReturnImmediatelyIfRetryIntervalIsGreaterThanOperationTimeout()
+        {
+            var policy = RetryPolicy.Default;
+            Stopwatch watch = Stopwatch.StartNew();
+            try
+            {
+                await policy.RunOperation(() => { throw new ServiceBusException(true, string.Empty); }, TimeSpan.FromSeconds(3));
+            }
+            catch (Exception)
+            {
+                // Expected
+            }
+            Assert.True(watch.Elapsed.TotalSeconds < 3);
+        }
+
+        [Fact]
+        async void RunOperationShouldWaitFor10SecondsForOperationIfServerBusy()
+        {
+            var policy = RetryPolicy.Default;
+            policy.SetServerBusy(Resources.DefaultServerBusyException);
+            Stopwatch watch = Stopwatch.StartNew();
+            try
+            {
+                await policy.RunOperation(() => Task.CompletedTask, TimeSpan.FromMinutes(3));
+            }
+            catch (Exception)
+            {
+                // Expected
+            }
+            Assert.True(watch.Elapsed.TotalSeconds > 9);
+            Assert.False(policy.IsServerBusy);
+        }
+
+        [Fact]
+        async void RunOperationShouldWaitForAllOperationsToSucceed()
+        {
+            var policy = RetryPolicy.Default;
+            Stopwatch watch = Stopwatch.StartNew();
+            try
+            {
+                await policy.RunOperation(
+                    async () =>
+                    {
+                        for (int i = 0; i < 5; i++)
```

Comment:
@SeanFeldman , True, i'm already taking care of it by using Fun<Task> instead of Action. But isn't that what automated test cases all about? Ensuring there is no regressions later? Ensuring that code isn't changed unknowingly next time without keeping this in mind?

And yes, ensuring tasks are in completed states is a valid approach. The current implementation also ensures that. Also ensures that multiple operations are not running in parallel which may happen with bad coding practices, which isn't too much to test, but an extra freebie is always good :)